### ISSUE TITLE
Public identifiers should have a prefix to avoid name clashes

### ISSUE THREAD

**Issue by ecorm** [ulfjack/ryu#148]
Public identifiers (extern functions and public enumerators) should have a prefix to avoid name collisions with user code and other libraries.

Extern functions should have a prefix such as `ryu_`.

The ryu_parse [Status enumerators](https://github.com/ulfjack/ryu/blob/447d6998bd3af17d675955186ab5b3cc3c585f00/ryu/ryu_parse.h#L32) should have a prefix such as `RYU_` or `RYU_STATUS_` to avoid clashing with macros as well as identifiers in the global namespace. The `SUCCESS` enumerator is particularly susceptible to conflict.

**Comment by expnkx** [comment]
Please no.

**Comment by ecorm** [comment]
@expnkx Why not? C doesn't have namespaces, so there is no other mechanism to prevent name clashes.

`d2s`, `f2s`, and `d2exp` are very short, and `SUCCESS` is very common. They are vulnerable to name clashes with user C code and other C libraries.

**Comment by expnkx** [comment]
> @expnkx Why not? C doesn't have namespaces, so there is no other mechanism to prevent name clashes.
> 
> `d2s`, `f2s`, and `d2exp` are very short, and `SUCCESS` is very common. They are vulnerable to name clashes with user C code and other C libraries.

You can clone that and put it into the namespace.

**Comment by ecorm** [comment]
> You can clone that and put it into the namespace. It all depends on you.

That presumes the user of this library is using C++ and not C. It also makes it impossible to use the existing Bazel script to build this library (compiled symbols will not be desired namespace).

I don't want to have to fork this library to make it usable in a large project where name clashes are likely.

> BTW, shared linking seriously harms the performance and security of computer.

The library can be statically linked, if desired. I doubt inlining the public library functions makes a significant performance difference versus static linking. These aren't small functions we're talking out.

**Comment by expnkx** [comment]
> The library can be statically linked, if desired. I doubt inlining the public library functions makes a significant performance difference versus static linking. These aren't small functions we're talking out.

They do not use it in a large library. Is that simple?

"I doubt inlining the public library functions makes a significant performance difference versus static linking. "

TBH, very significant and that is why my library outperforms ryu source for at least 20% on GCC.

**Comment by expnkx** [comment]
Like I said before if you want a good interface and performance. NEVER use any C APIs

**Comment by expnkx** [comment]
BTW. I do not know what you are going to do with the result of these algorithms like ryu. Are you going to write to a file or network or something? That means ryu won't save you at all since stream and cstdio are both too slow which eliminates all the benefit you come from ryu.

**Comment by ulfjack** [comment]
Yeah, they should probably be prefixed. The code is now part of MSVC and (soon?) libc++, and gets prefixed on import there. However, if you want to use it directly, it's a bit annoying that it can cause name conflicts as-is.

**Comment by ecorm** [comment]
My motivation is to use ryu as a substitute for `std::from_chars` and `std::to_chars` on compilers that don't support them.

**Comment by ecorm** [comment]
I don't know if the goal of this project is to provide a library for general use, or to serve as a reference for other Ryu implementations (or both). Please let know.

I could hide the ryu symbols within a separately compiled unit, but that would preclude the possibility of making my C++ library header-only. For better or worse, header-only C++ libraries are popular due to their ease of integration. Mine would be optionally header-only, and could be compiled separately if desired.

To avoid breaking the existing API, you could provide alternate headers that have the prefixed versions of functions and other symbols. The prefixed versions would simply delegate to the non-prefixed versions. The extra indirection should be optimized away.

**Comment by ulfjack** [comment]
I'm ok with breaking the existing API.
